test/Microsoft.AspNetCore.Razor.LanguageServer.Test/DefaultTagHelperCompletionServiceTest.cs

```diff
@@ -236,6 +261,200 @@ public void GetCompletionAt_AtTagHelperElementNameEdge_ReturnsCompletions()
+            var service = new DefaultTagHelperCompletionService(RazorTagHelperCompletionService);
+            var codeDocument = CreateCodeDocument($"@addTagHelper *, TestAssembly{Environment.NewLine}<test2 unbound />", DefaultTagHelpers);
+            var sourceSpan = new SourceSpan(38 + Environment.NewLine.Length, 0);
+
+            // Act
+            var completions = service.GetCompletionsAt(sourceSpan, codeDocument);
+
+            // Assert
+            Assert.Collection(
+                completions,
+                completion => Assert.Equal("bool-val", completion.FilterText),
+                completion => Assert.Equal("int-val", completion.FilterText));
+        }
+
+        [Fact]
+        public void GetCompletionAt_MinimizedTagHelperAttribute_ReturnsCompletions()
+        {
+            // Arrange
+            var service = new DefaultTagHelperCompletionService(RazorTagHelperCompletionService);
+            var codeDocument = CreateCodeDocument($"@addTagHelper *, TestAssembly{Environment.NewLine}<test2 bool-val />", DefaultTagHelpers);
+            var sourceSpan = new SourceSpan(38 + Environment.NewLine.Length, 0);
+
+            // Act
+            var completions = service.GetCompletionsAt(sourceSpan, codeDocument);
+
+            // Assert
+            Assert.Collection(
+                completions,
+                completion => Assert.Equal("int-val", completion.FilterText));
+        }
+
+        [Fact]
+        public void GetCompletionAt_HtmlAttribute_ReturnsCompletions()
+        {
+            // Arrange
+            var service = new DefaultTagHelperCompletionService(RazorTagHelperCompletionService);
+            var codeDocument = CreateCodeDocument($"@addTagHelper *, TestAssembly{Environment.NewLine}<test2 class='' />", DefaultTagHelpers);
+            var sourceSpan = new SourceSpan(38 + Environment.NewLine.Length, 0);
+
+            // Act
+            var completions = service.GetCompletionsAt(sourceSpan, codeDocument);
+
+            // Assert
+            Assert.Collection(
+                completions,
+                completion => Assert.Equal("bool-val", completion.FilterText),
+                completion => Assert.Equal("int-val", completion.FilterText));
+        }
+
+        [Fact]
+        public void GetCompletionAt_TagHelperAttribute_ReturnsCompletions()
+        {
+            // Arrange
+            var service = new DefaultTagHelperCompletionService(RazorTagHelperCompletionService);
+            var codeDocument = CreateCodeDocument($"@addTagHelper *, TestAssembly{Environment.NewLine}<test2 int-val='123' />", DefaultTagHelpers);
+            var sourceSpan = new SourceSpan(38 + Environment.NewLine.Length, 0);
+
+            // Act
+            var completions = service.GetCompletionsAt(sourceSpan, codeDocument);
+
+            // Assert
+            Assert.Collection(
+                completions,
+                completion => Assert.Equal("bool-val", completion.FilterText));
+        }
```

Comment:
I suggest adding some malformed attribute cases. Maybe even another element inside an attribute value. These cases tend to be pretty common while typing.

Reply:
It's interesting, additional elements inside of attribute values aren't parsed as elements. Instead they look like:

```CSHTML
@addTagHelper *, TestAssembly
<test2 string-val=\" @<test2 int-val='123' />\"
```

```
RazorDocument - [0..76) - FullWidth: 76 - [@addTagHelper *, TestAssemblyLF<test2 string-val=" @<test2 int-val='123' />"]
    MarkupBlock - [0..76) - FullWidth: 76
        MarkupTextLiteral - [0..0) - FullWidth: 0 - [] - Gen<Markup> - SpanEditHandler;Accepts:Any
            Marker;[];
        CSharpCodeBlock - [0..29) - FullWidth: 29
            RazorDirective - [0..29) - FullWidth: 29
                CSharpTransition - [0..1) - FullWidth: 1 - Gen<None> - SpanEditHandler;Accepts:None
                    Transition;[@];
                RazorDirectiveBody - [1..29) - FullWidth: 28
                    RazorMetaCode - [1..13) - FullWidth: 12 - Gen<None> - SpanEditHandler;Accepts:None
                        Identifier;[addTagHelper];
                    CSharpCodeBlock - [13..29) - FullWidth: 16
                        MarkupTextLiteral - [13..14) - FullWidth: 1 - [ ] - Gen<None> - SpanEditHandler;Accepts:None
                            Whitespace;[ ];
                        CSharpStatementLiteral - [14..29) - FullWidth: 15 - [*, TestAssembly] - Gen<AddTagHelper:{*, TestAssembly;*, TestAssembly;*;TestAssembly}> - SpanEditHandler;Accepts:AnyExceptNewline
                            Star;[*];
                            Comma;[,];
                            Whitespace;[ ];
                            Identifier;[TestAssembly];
        MarkupTextLiteral - [29..31) - FullWidth: 2 - [LF] - Gen<Markup> - SpanEditHandler;Accepts:Any
            NewLine;[LF];
        MarkupTagHelperElement - [31..76) - FullWidth: 45 - test2[StartTagAndEndTag] - TestTagHelper
            MarkupTagHelperStartTag - [31..76) - FullWidth: 45 - [<test2 string-val=" @<test2 int-val='123' />"] - Gen<Markup> - SpanEditHandler;Accepts:Any
                OpenAngle;[<];
                Text;[test2];
                MarkupTagHelperAttribute - [37..76) - FullWidth: 39 - string-val - DoubleQuotes - Bound - [ string-val=" @<test2 int-val='123' />"]
                    MarkupTextLiteral - [37..38) - FullWidth: 1 - [ ] - Gen<Markup> - SpanEditHandler;Accepts:Any
                        Whitespace;[ ];
                    MarkupTextLiteral - [38..48) - FullWidth: 10 - [string-val] - Gen<Markup> - SpanEditHandler;Accepts:Any
                        Text;[string-val];
                    Equals;[=];
                    MarkupTextLiteral - [49..50) - FullWidth: 1 - ["] - Gen<None> - SpanEditHandler;Accepts:Any
                        DoubleQuote;["];
                    MarkupTagHelperAttributeValue - [50..75) - FullWidth: 25
                        MarkupBlock - [50..52) - FullWidth: 2
                            MarkupTextLiteral - [50..51) - FullWidth: 1 - [ ] - Gen<Markup> - SpanEditHandler;Accepts:Any
                                Whitespace;[ ];
                            CSharpCodeBlock - [51..52) - FullWidth: 1
                                CSharpImplicitExpression - [51..52) - FullWidth: 1
                                    CSharpTransition - [51..52) - FullWidth: 1 - Gen<None> - SpanEditHandler;Accepts:None
                                        Transition;[@];
                                    CSharpImplicitExpressionBody - [52..52) - FullWidth: 0
                                        CSharpCodeBlock - [52..52) - FullWidth: 0
                                            CSharpExpressionLiteral - [52..52) - FullWidth: 0 - [] - Gen<Expr> - ImplicitExpressionEditHandler;Accepts:NonWhitespace;ImplicitExpression[RTD];K17
                                                Marker;[];
                        MarkupTextLiteral - [52..58) - FullWidth: 6 - [<test2] - Gen<Markup> - SpanEditHandler;Accepts:Any
                            OpenAngle;[<];
                            Text;[test2];
                        MarkupTextLiteral - [58..72) - FullWidth: 14 - [ int-val='123'] - Gen<Markup> - SpanEditHandler;Accepts:Any
                            Whitespace;[ ];
                            Text;[int-val];
                            Equals;[=];
                            SingleQuote;['];
                            Text;[123];
                            SingleQuote;['];
                        MarkupTextLiteral - [72..75) - FullWidth: 3 - [ />] - Gen<Markup> - SpanEditHandler;Accepts:Any
                            Whitespace;[ ];
                            ForwardSlash;[/];
                            CloseAngle;[>];
                    MarkupTextLiteral - [75..76) - FullWidth: 1 - ["] - Gen<None> - SpanEditHandler;Accepts:Any
                        DoubleQuote;["];
                CloseAngle;[<Missing>];
```

WIll skip on the attributes inside of attributes test but will add other malformed cases.